projects/plugins/crm/src/automation/commons/actions/contacts/class-update-contact.php

```diff
@@ -73,33 +76,27 @@ public static function get_category(): ?string {
 		return __( 'Contact', 'zero-bs-crm' );
 	}
 
-	/**
-	 * Get the allowed triggers.
-	 *
-	 * @since $$next-version$$
-	 *
-	 * @return string[]|null The allowed triggers.
-	 */
-	public static function get_allowed_triggers(): ?array {
-		return array();
-	}
-
 	/**
 	 * Update the DAL with the new contact data.
 	 *
 	 * @since $$next-version$$
 	 *
-	 * @param mixed  $data Data passed from the trigger.
-	 * @param ?mixed $previous_data (Optional) The data before being changed.
+	 * @param Data_Type $data Data passed from the trigger.
 	 */
-	public function execute( $data, $previous_data = null ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+	public function execute( Data_Type $data ) {
+		$this->validate( $data );
```

Comment:
It's not blocking, so we shouldn't address it in this specific PR (we can do it in a follow-up if we feel like it's "critical"), but it would be awesome if we didn't have to do `$this->validate` in each step execute method.

**Suggestion:** We could move the logic to `Automation_Engine::execute_workflow` and check `validate()` before we run the `execute()` method.

**Why?**: It will make it more intuitive for 3. party developers to create steps.

Reply:
Yes, absolutely. I felt the same while I was adding it, but not energy to stop and think about how to do it and move it elsewhere. Moving it into `Automation_Engine::execute_workflow` sounds good.